okhttp/src/main/java/okhttp3/internal/http2/Hpack.kt

```diff
@@ -13,576 +13,584 @@
+    @Throws(IOException::class)
+    private fun readByte(): Int {
+      return source.readByte() and 0xff
     }
 
-    int readInt(int firstByte, int prefixMask) throws IOException {
-      int prefix = firstByte & prefixMask;
+    @Throws(IOException::class)
+    fun readInt(firstByte: Int, prefixMask: Int): Int {
+      val prefix = firstByte and prefixMask
       if (prefix < prefixMask) {
-        return prefix; // This was a single byte value.
+        return prefix // This was a single byte value.
       }
 
       // This is a multibyte value. Read 7 bits at a time.
-      int result = prefixMask;
-      int shift = 0;
+      var result = prefixMask
+      var shift = 0
       while (true) {
-        int b = readByte();
-        if ((b & 0x80) != 0) { // Equivalent to (b >= 128) since b is in [0..255].
-          result += (b & 0x7f) << shift;
-          shift += 7;
+        val b = readByte()
+        if (b and 0x80 != 0) { // Equivalent to (b >= 128) since b is in [0..255].
+          result += b and 0x7f shl shift
+          shift += 7
         } else {
-          result += b << shift; // Last byte.
-          break;
+          result += b shl shift // Last byte.
+          break
         }
       }
-      return result;
+      return result
     }
 
-    /** Reads a potentially Huffman encoded byte string. */
-    ByteString readByteString() throws IOException {
-      int firstByte = readByte();
-      boolean huffmanDecode = (firstByte & 0x80) == 0x80; // 1NNNNNNN
-      int length = readInt(firstByte, PREFIX_7_BITS);
+    /** Reads a potentially Huffman encoded byte string.  */
+    @Throws(IOException::class)
+    fun readByteString(): ByteString {
+      val firstByte = readByte()
+      val huffmanDecode = firstByte and 0x80 == 0x80 // 1NNNNNNN
+      val length = readInt(firstByte, PREFIX_7_BITS)
 
       if (huffmanDecode) {
-        return ByteString.of(Huffman.get().decode(source.readByteArray(length)));
+        val bytes = source.readByteArray(length.toLong())
+        return ByteString.of(*Huffman.get().decode(bytes))
       } else {
-        return source.readByteString(length);
+        return source.readByteString(length.toLong())
       }
     }
   }
 
-  static final Map<ByteString, Integer> NAME_TO_FIRST_INDEX = nameToFirstIndex();
-
-  private static Map<ByteString, Integer> nameToFirstIndex() {
-    Map<ByteString, Integer> result = new LinkedHashMap<>(STATIC_HEADER_TABLE.length);
-    for (int i = 0; i < STATIC_HEADER_TABLE.length; i++) {
+  private fun nameToFirstIndex(): Map<ByteString, Int> {
+    val result = LinkedHashMap<ByteString, Int>(STATIC_HEADER_TABLE.size)
+    for (i in STATIC_HEADER_TABLE.indices) {
       if (!result.containsKey(STATIC_HEADER_TABLE[i].name)) {
-        result.put(STATIC_HEADER_TABLE[i].name, i);
+        result[STATIC_HEADER_TABLE[i].name] = i
       }
     }
-    return Collections.unmodifiableMap(result);
+    return Collections.unmodifiableMap(result)
```

Comment:
Is this still needed here, or is immutable Map vs MutableMap strong enough here?

Reply:
I don’t like Kotlin’s immutable types. They aren’t immutable to Java callers, and they aren’t immutable if you cast.

Reply:
Not sure we should care about casts. Once you go down that road you can always get access to mutable internals.

Reply:
I suppose it’s okay for internal APIs.

For published APIs returning a mutable result discourages us from returning an immutable result later.

It'd be much less problematic if Java’s type system disambiguated the two.